src/lib/reasoners/relation.ml

```diff
@@ -122,23 +122,40 @@ let query env uf a =
             | Some _ as ans -> ans
             | None -> Rel7.query env.r7 uf a
 
-let case_split env uf ~for_model =
+let case_split env uf ~for_model ~to_optimize =
   Options.exec_thread_yield ();
-  let seq1 = Rel1.case_split env.r1 uf ~for_model in
-  let seq2 = Rel2.case_split env.r2 uf ~for_model in
-  let seq3 = Rel3.case_split env.r3 uf ~for_model in
-  let seq4 = Rel4.case_split env.r4 uf ~for_model in
-  let seq5 = Rel5.case_split env.r5 uf ~for_model in
-  let seq6 = Rel6.case_split env.r6 uf ~for_model in
-  let seq7 = Rel7.case_split env.r7 uf ~for_model in
-  let l = seq1 |@| seq2 |@| seq3 |@| seq4 |@| seq5 |@| seq6 |@| seq7 in
-  List.sort
-    (fun (_,_,sz1) (_,_,sz2) ->
-       match sz1, sz2 with
-       | Th_util.CS(_,sz1), Th_util.CS(_,sz2) -> Numbers.Q.compare sz1 sz2
-       | _ -> assert false
-    )l
-
+  let seq1 = Rel1.case_split env.r1 uf ~for_model ~to_optimize in
+  let seq2 = Rel2.case_split env.r2 uf ~for_model ~to_optimize in
+  let seq3 = Rel3.case_split env.r3 uf ~for_model ~to_optimize in
+  let seq4 = Rel4.case_split env.r4 uf ~for_model ~to_optimize in
+  let seq5 = Rel5.case_split env.r5 uf ~for_model ~to_optimize in
+  let seq6 = Rel6.case_split env.r6 uf ~for_model ~to_optimize in
+  let seq7 = Rel7.case_split env.r7 uf ~for_model ~to_optimize in
```

Comment:
I suggest to do it in a separated PR.

Reply:
I don't know — I think we can do it here. You have done a great job over the last couple months maintaining the branch as some of the fixes were also made on the main branch, and then splitting what was once a huge PR into three self-contained and reasonably sized PRs (satML models, all-sat, and optimization). And since the PR is now feature-sized, we can evaluate it on its own, and I think it is reasonable to clean it up before merging. I can take that up if you're busy elsewhere.

Reply:
Ok, I will do it now.